Solve_constraints and Set Use Unification Heuristics

mattam82 opened the pull request:
This follows PR #330 , exporting solve_constraints as a tactic and Unset Use Unification Heuristics (it is Set by default), allowing to play with unification heuristics from the Ltac side. This is highly experimental and could very well be removed from the final, but I feel we should provide a way for our users to play with it and don't see how to do this from a plugin.

mattam82 commented:
Also, the code to print unification constraints and candidates becomes useless without this :)

mattam82 commented:
https://ci.inria.fr/coq/view/coq-contribs/job/coq-contribs/520/console

aspiwack commented:
As a general remark: even if a PR expands documentation (which you should eventually do before merging I suppose) it is a good practice to include some of that documentation in the PR exposé. Otherwise most concerned people will probably not understand what the PR is about.

Digging through the issue linked in the linked PR, I was able to figure out more or less the missing piece. Relevant quote:

> Adds a new multi-goal tactic [solve_constraints] that forces solving of
unification constraints and evar candidates to be solved.
Adds an option [Set Eager Tactic Unification] that launches
[solve_constraints] at every [.], on by default, preserving the 8.4
behavior of tactics. 

It doesn't seem to be quite up to date though, is it?

mattam82 commented:
No it's not, I have to rework this a little bit before it can be merged actually. Sorry for being obscure, I assume the commit messages are part of the PR "exposé" usually.

mattam82 commented:
Rebased, moved code associated with PR #330 here, add flag in Compat/Coq85 and do renamings of consider_remaining and reconsider_conv_pbs to be more uniform and explicit. This solves bugs #5149 and #4763

mattam82 commented:
https://ci.inria.fr/coq/view/coq-contribs/job/coq-contribs/543/console

maximedenes commented:
If no objection, I will rename `Set Use Unification Heuristics` into `Set Solve Unification Constraints`. And similarly, in the pretyping API, `inference_flags` should have a field named `solve_unification_constraints`, rather than `use_heuristics`. In particular, setting it to `true` can trigger more failures.

herbelin commented:
Sorry, I did not follow, is this just renaming of the field or a change of semantics?

About the renaming of the option, can you remember the motivation?

maximedenes commented:
It is just a renaming, no change of semantics. The reason is that I find misleading that `Set Use Unification Heuristics` does not change the way unification uses heuristics. Typically, you turn this option on when unification use *fewer* heuristics, so that the "." triggers them. Also, I wouldn't expect an option called `Set Use Unification Heuristics` to fail more often when on. In fact what this option is doing is to make sure we solve remaining constraints at each tactic end point (".").

maximedenes commented:
Not a big deal though, if people prefer the current name, we can leave it as it is.

mattam82 commented:
Yes, the name makes more sense indeed.

On Fri, Oct 28, 2016 at 2:14 PM Maxime Dénès <notifications@github.com>
wrote:

> It is just a renaming, no change of semantics. The reason is that I find
> misleading that Set Use Unification Heuristics does not change the way
> unification uses heuristics. Typically, you turn this option on when
> unification use *fewer* heuristics, so that the "." triggers them. Also,
> I wouldn't expect an option called Set Use Unification Heuristics to fail
> more often when on. In fact what this option is doing is to make sure we
> solve remaining constraints at each tactic end point (".").
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/coq/coq/pull/331#issuecomment-256904635>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/AAGARaN43_GfyX57m7kHr_z858ph9Qh9ks5q4eY6gaJpZM4KcLDb>
> .
>

herbelin commented:
For "Set Solve Unification Constraints", I was a bit concerned by the use of the word "Solve" because not only it may fail (hence not solve) but when succeeding, the solution is a priori not the but only a solution. 

On the other side, for "Set Use Unification Heuristics", one of the problem is that unification heuristics are used anyway in some part of the unification algorithm even when the option is unset.

So, this is a bit long, but you'd like to express something like "Set Ensure No Pending Unification Constraints At Completion Of The Tactic Sentence", right? If yes, we would need to shorten that ("Set No Pending Constraints"? or, turned the other way round "Unset Accept Pending Constraints"?)

About the field, it seems to me that use_heuristics remain relevant (for the reasons above that it may not "solve" them in the strict sense: it may fail or take arbitrary decisions). But maybe you would prefer e.g. try_unification_heuristics to highlight further that it may fail?

maximedenes commented:
I'm a bit confused here: if unification cannot solve these constraints, the function fails, right? But `try_unification_heuristics` conveys exactly the opposite idea, that it will try without failing if unification cannot solve the constraints.

maximedenes commented:
@herbelin You are right, I was confused. The flag in `inference_flags` indeed allows pending constraints, so its name is fine. However, the fact that it is always called with value `true` seems to indicate that there will be more work to be done to decouple unification (without heuristics) and tactics that implement heuristics (like Matthieu's `solve_constraints`).

herbelin commented:
> I'm a bit confused here: if unification cannot solve these constraints, the function fails, right?

Yes (unless fail_evar is false).

Otherwise said:

> The flag in inference_flags indeed allows pending constraints, 

No (unless fail_evar is false).

> But try_unification_heuristics conveys exactly the opposite idea, that it will try without failing if unification cannot solve the constraints.

Yes, I agree that "try" is ambiguous, in the natural language meaning I think it is ok, but in the OCaml meaning, it suggests that failures are caught. So, let's renounce to using "try".

> ... there will be more work to be done to decouple unification (without heuristics) and tactics that implement heuristics (like Matthieu's solve_constraints).

Yes!

maximedenes commented:
We need to wrap up this discussion. I believe we agreed that `solve_unification_constraints` was an ok name in `inference_flags`.

Now, about the option name, I tend to prefer `Set Solve Unification Constraints` rather than `Set No Pending Constraints` because this option does not simply forbid such constraints, it also tries to solve it (at each "."). Also, I'm not sure to there is a problem in using the world "solve" even when there might be other solutions.

mattam82 commented:
LGTM

herbelin commented:
I'm still not convinced that `Set Solve Unification Constraints` is clearer. Even without this option, unification heuristics are sometimes used at other parts of the unification path. What the option does is that it ensures that at the `.`, a failure is given if the algorithm trying heuristics is still not able to solve all the constraints still existing at this time.

mattam82 commented:
It's not about using heuristics (we always do) but about solving the
constraints then, so I don't see the problem.

On Fri, Nov 4, 2016 at 4:42 PM Hugo Herbelin <notifications@github.com>
wrote:

> I'm still not convinced that Set Solve Unification Constraints is
> clearer. Even without this option, unification heuristics are sometimes
> used at other parts of the unification path. What the option does is that
> it ensures that at the ., a failure is given if the algorithm trying
> heuristics is still not able to solve all the constraints still existing at
> this time.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/coq/coq/pull/331#issuecomment-258466592>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/AAGARTI50JGEiyJklq0RF1dlj7d27F28ks5q61JcgaJpZM4KcLDb>
> .
>

herbelin commented:
From the practical point of view, we do need a name and I have obviously no more consensual alternative to propose, so the proposed name is ok to me.
